src/layers/trip-layer/trip-utils.js

```diff
@@ -0,0 +1,188 @@
+
+  if (!analyzedType || analyzedType.category !== 'TIME') {
+    return false;
+  }
+  return analyzedType;
+}
+
+/**
+ * Check if geojson features are trip layer animatable by meeting 3 conditions
+ * @param {array} features array of geojson feature objects
+ * @returns {boolean} whether it is trip layer animatable
+ */
+export function isTripGeoJsonField(allData = [], field) {
+  if (!allData.length) {
+    return false;
+  }
+  const getValue = d => d[field.tableFieldIndex - 1];
+  const maxCount = 10000;
+  const sampleRawFeatures =
+    allData.length > maxCount ? getSampleData(allData, maxCount, getValue) : allData.map(getValue);
+
+  const features = sampleRawFeatures.map(parseGeoJsonRawFeature);
+
+  let isTrip = false;
+  const featureTypes = getGeojsonFeatureTypes(features);
+
+  // condition 1: contain line string
+  const hasLineString = Boolean(featureTypes.line);
+  if (!hasLineString) {
+    return isTrip;
+  }
+
+  // condition 2:sample line strings contain 4 coordinates
+  const HasLength4 = coordHasLength4(features);
+  if (!HasLength4) {
+    return isTrip;
+  }
+
+  // condition 3:the 4th coordinate of the first feature line strings is valid time
+  const tsHolder = features[0].geometry.coordinates.map(
+    coord => coord[3]
+  );
+
+  const hasValidTime = containValidTime(tsHolder);
+  if (hasValidTime) {
+    isTrip = true;
+  }
+
+  return isTrip;
+}
+
+/**
+ * Get unix timestamp from animatable geojson for deck.gl trip layer
+ * @param {Array<Object>} dataToFeature array of geojson feature objects, can be null
+ * @returns {Array<Number>} unix timestamp in milliseconds
+ */
+export function parseTripGeoJsonTimestamp(dataToFeature) {
+  // Analyze type based on coordinates of the 1st lineString
+  // select a sample trip to analyze time format
+  const empty = {dataToTimeStamp: [], animationDomain: null};
+  const sampleTrip = dataToFeature.find(
+    f =>
+      f &&
+      f.geometry &&
+      f.geometry.coordinates &&
+      f.geometry.coordinates.length >= 3
+  );
+
+  // if no valid geometry
+  if (!sampleTrip) {
+    return empty;
+  }
+
+  const analyzedType = containValidTime(
+    sampleTrip.geometry.coordinates.map(coord => coord[3])
+  );
+
+  if (!analyzedType) {
+    return empty;
+  }
+
+  const {format} = analyzedType;
+  const getTimeValue = coord =>
+    coord && notNullorUndefined(coord[3])
+      ? timeToUnixMilli(coord[3], format)
+      : null;
+
+  const dataToTimeStamp = dataToFeature.map(f =>
+    f && f.geometry && Array.isArray(f.geometry.coordinates)
+      ? f.geometry.coordinates.map(getTimeValue)
+      : null
+  );
+
+  const animationDomain = getAnimationDomainFromTimestamps(dataToTimeStamp);
+
+  return {dataToTimeStamp, animationDomain};
+}
+
+function findMinFromSorted(list = []) {
+  let i = 0;
+  while (i < list.length) {
+    if (notNullorUndefined(list[i])) {
+      return list[i]
+    }
+    i++;
+  }
+  return null;
+}
+
+function findMaxFromSorted(list = []) {
```

Comment:
here you can take advantage of inverse:
```javascript
return list.reverse().find(value => notNullorUndefined(value));
```

Reply:
the list can be really long, when it's sorted, starting from the end  and immediately return when found

Reply:
actually i was reading some performance reports and it looks liek reverse is not as performant as doing a while loop. Good with this